lib/Runtime/Types/PathTypeHandler.h

```diff
@@ -190,20 +245,38 @@ namespace Js
         BOOL ConvertToSharedNonExtensibleTypeIfNeededAndCallOperation(DynamicObject* instance, const PropertyRecord* operationInternalPropertyId, FType operation);
 
         template <bool isObjectLiteral>
-        DynamicType* PromoteType(DynamicType* type, const PropertyRecord* propertyId, bool shareType, ScriptContext* scriptContext, DynamicObject* object = nullptr, PropertyIndex* propertyIndex = nullptr);
+        DynamicType* PromoteType(DynamicType* type, const PathTypeSuccessorKey key, bool shareType, ScriptContext* scriptContext, DynamicObject* object = nullptr, PropertyIndex* propertyIndex = nullptr);
+        ObjectSlotAttributes * UpdateAttributes(Recycler * recycler, ObjectSlotAttributes * oldAttributes, uint8 oldPathSize, TypePath * newTypePath);
+        SetterSlotIndex * UpdateSetterSlots(Recycler * recycler, SetterSlotIndex * oldSetters, uint8 oldPathSize, TypePath * newTypePath);
 
         PropertyIndex GetPropertyIndex(PropertyId propertyId);
 
         void SetSlotAndCache(DynamicObject* instance, PropertyId propertyId, PropertyRecord const * record, PropertyIndex index, Var value, PropertyValueInfo* info, PropertyOperationFlags flags, SideEffects possibleSideEffects);
     protected:
-        virtual bool GetSuccessor(const PropertyRecord* propertyRecord, RecyclerWeakReference<DynamicType> ** typeWeakRef) = 0;
-        virtual void SetSuccessor(DynamicType * type, const PropertyRecord* propertyRecord, RecyclerWeakReference<DynamicType> * typeWeakRef, ScriptContext * scriptContext) = 0;
+        virtual bool GetSuccessor(const PathTypeSuccessorKey successorKey, RecyclerWeakReference<DynamicType> ** typeWeakRef) = 0;
+        virtual void SetSuccessor(DynamicType * type, const PathTypeSuccessorKey successorKey, RecyclerWeakReference<DynamicType> * typeWeakRef, ScriptContext * scriptContext) = 0;
 
         uint16 GetPathLength() const { return GetUnusedBytesValue(); }
         TypePath * GetTypePath() const { return typePath; }
+        virtual void SetTypePath(TypePath *typePath) { this->typePath = typePath; }
         DynamicType * GetPredecessorType() const { return predecessorType; }
         PathTypeHandlerBase* GetRootPathTypeHandler();
 
+        virtual ObjectSlotAttributes * GetAttributeArray() const { return nullptr; }
+        virtual ObjectSlotAttributes GetAttributeArray(const PropertyIndex index) const { return ObjectSlotAttr_Default; }
```

Comment:
>GetAttributeArray [](start = 37, length = 17)

This is returning the attributes for a particular property index, right? If so, can we rename it to GetAttributes? 

Proposing similar change for GetSetterSlots below.

Reply:
Oh, right. That's what I had originally; didn't mean to change it. I'll change it back.